### PR TITLE
Define countries for a channel

### LINKED ISSUE #9122: Countries available in the checkout made configurable per channel
⚠️ 🛣 **This issue is a part of the** [Sylius Public Roadmap](https://github.com/orgs/Sylius/projects/2) 🛣 ⚠️

| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Sylius version   | 1.2

Currently, if we have 2 channels "US Webstore" and "EU Webstore", during the checkout, all countries are available. An improvement would be to add a relation between every Channel and many Countries. And if Channel has like 2 out of 5 countries selected, these are available during checkout. If no countries are selected - all are available. So we retain the current behavior (no BC break) but add new, useful feature.

To achieve this, I would suggest to modify the AddressType form type. In the CheckoutAddressType form type we could identify the channel of the order and use appropriate countries as options to choose from.

### PR DESCRIPTION
**Pull request by pamil**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #9122 
| License         | MIT

Currently used only in checkout addressing to limit the list of countries. If not defined, fallbacks to the previous behaviour - all enabled countries.

Screenshot 2020-01-22 at 20 10 51

### CODE REVIEW

File: features/checkout/addressing_order/restricting_list_of_countries_available_for_addressing.feature [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        And the store operates in "Poland"
+        And the store has a product "PHP T-Shirt" priced at "$19.99"
+        And the store ships everywhere for free
+        And I am a logged in customer
+
+    @ui
+    Scenario: Having only countries available for current channel listed
+        Given this channel operates in the "United States" country
+        When I add product "PHP T-Shirt" to the cart
+        And I am at the checkout addressing step
```
Comment (GSadee) [inline features/checkout/addressing_order/restricting_list_of_countries_available_for_addressing.feature]:
```suggestion
        And I go to the checkout addressing step
```
?

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 2]
```diff
@@ -38,21 +47,13 @@ public function configureOptions(OptionsResolver $resolver): void
         $resolver
             ->setDefaults([
                 'choice_filter' => null,
-                'choices' => function (Options $options): iterable {
+                'choices' => function (Options $options, $prevValue): iterable {
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php]: Is that `$prevValue` needed?

File: src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php [hunk 3]
```diff
@@ -63,6 +64,17 @@ public function configureOptions(OptionsResolver $resolver): void
                 'placeholder' => 'sylius.form.country.select',
             ])
             ->setAllowedTypes('choice_filter', ['null', 'callable'])
+            ->setNormalizer('choices', static function (Options $options, array $countries): array {
+                if ($options['choice_filter']) {
+                    $countries = array_filter($countries, $options['choice_filter']);
+                }
+
+                usort($countries, static function (CountryInterface $a, CountryInterface $b): int {
```
Comment (GSadee) [inline src/Sylius/Bundle/AddressingBundle/Form/Type/CountryChoiceType.php]:
```suggestion
                usort($countries, static function (CountryInterface $firstCountry, CountryInterface $secondCountry): int {
```
?

File: src/Sylius/Behat/Page/Shop/Checkout/AddressPage.php [hunk 4]
```diff
@@ -283,6 +293,18 @@ protected function getDefinedElements(): array
...
 
+    /**
+     * @return string[]
+     */
+    private function getOptionsFromSelect(NodeElement $element): array
+    {
+        return array_map(
+            /** @return string[] */
+            static function (NodeElement $element): string { return $element->getText(); },
+            $element->findAll('css', 'option[value!=""]')
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/Checkout/AddressPage.php]: I'm wondering if we cannot support data attributes here as well

**Comment by lchrusciel** [comment]
Thank you, Kamil! :1st_place_medal:

**Pull request merged**